go/vt/vtgate/planbuilder/builder.go

```diff
@@ -119,6 +120,71 @@ func BuildFromStmt(ctx context.Context, query string, stmt sqlparser.Statement,
 	return plan, nil
 }
 
+func buildMirrorPlan(ctx context.Context, from *planResult, query string, stmt sqlparser.Statement, reservedVars *sqlparser.ReservedVars, vschema plancontext.VSchema, enableOnlineDDL, enableDirectDDL bool) (*planResult, error) {
+	// Collect keyspaces and validate tables.
+	fromKeyspaces := make(map[string]any)
+	toKeyspaces := make(map[string]float32)
+
+	for _, table := range from.tables {
+		// Record the from keyspace.
+		fromKeyspace := table.Qualifier.String()
+		fromKeyspaces[fromKeyspace] = true
+
+		// Find the mirror rule for the from keyspace.
+		mirrorRule, err := vschema.FindMirrorRule(fromKeyspace)
+		if err != nil {
+			return nil, err
+		}
+		if mirrorRule == nil {
+			return nil, nil
+		}
+
+		// Record the to keyspace.
+		toKeyspaces[mirrorRule.ToKeyspace] = mirrorRule.Percent
+
+		// Validate that the to table exists in the vschema (always true if the
+		// keyspace is unsharded).
+		table, err := vschema.GetVSchema().FindTable(mirrorRule.ToKeyspace, table.Name.String())
+		if err != nil {
+			return nil, err
+		}
+		if table == nil {
+			return nil, nil
+		}
+	}
+
+	// Nothing to mirror if we can't identify any to keyspaces.
+	if len(toKeyspaces) == 0 {
+		return nil, nil
+	}
+
+	// Use the smallest mirror percent.
+	var i int
+	var percent float32
+	for _, p := range toKeyspaces {
+		if i == 0 || p < percent {
+			percent = p
+		}
+		i++
+	}
+
+	mirrorVSchema := plancontext.ForMirroring(vschema)
+	target, err := createInstructionFor(ctx, query, stmt, reservedVars, mirrorVSchema, enableOnlineDDL, enableDirectDDL)
+	if err != nil {
+		return nil, err
+	}
+
+	target.primitive = engine.NewMirror(
+		from.primitive,
+		engine.NewPercentMirrorTarget(percent, target.primitive),
+	)
+	tables := append(from.tables, target.tables...)
+	operators.SortTableNames(tables)
```

Comment:
AFAICT we only need table names to be sorted for logging/stats purposes, e.g. so that `planbuilder` tests are consistent. Maybe we can delay sorting until it's needed?